File: decrypt.cpp
#include <iostream>
#include <windows.h>
#include <direct.h>
#include "cryptor.h"
#include "cryptor.cpp"

int main(int argc, char *argv[]){
    if (argc < 2) {
        std::cerr << "Wrong usage, type ./decrypt --help for more info." << std::endl;
        return 1;
    }

    if (strcmp(argv[1], "--help") == 0) {
        decryptHelp();
        return 0;
    }

    std::string inputFile, outputFile, keyInput, key, cwd = "";

    if (strcmp(argv[1], "-f") == 0 && argc == 4) {
        inputFile = argv[2];
        keyInput = argv[3];
    } else if (strcmp(argv[1], "-r") == 0) {
        cwd = getCurrentDir();
        keyInput = argv[2];
    } else if (argc == 3) {
        cwd = argv[1];
        keyInput = argv[2];
    } else {
        std::cerr << "Wrong usage, type ./decrypt --help for more info." << std::endl;
        return 1;
    }

    key = loadKey(keyInput);
    
    if (!cwd.empty()) {
        if (!Cryptor::DecryptRecur(key)){
            std::cerr << "Recursive decryption failed!" << std::endl;
            return 1;
        }
    } else {
        if (!Cryptor::DecryptFile(inputFile, key)) {
            std::cerr << "File decryption failed!" << std::endl;
            return 1;
        }
    }

    std::cout << "File(s) decrypted successfully";
    return 0;
}

File: decrypt.py
import sys, os
from enc_header import (
    decrypt_help,
    load_private_key,
    load_key,
    generate_asymmetric_keys,
    check_flags_dec,
    save_key_to_file,
    decrypt_directory,
    decrypt_file,
    decrypt_aes_key,
)
from cryptography.hazmat.primitives import serialization


def main():
    argc = len(sys.argv)
    argv = sys.argv

    if argc < 2:
        print("Wrong usage, type ./decrypt --help for more info.")
        sys.exit(1)

    if argv[1] == "--help":
        decrypt_help()
        return 0

    enc_type, inputFile, cwd, spec_dir, priv_key = check_flags_dec(argv, argc)

    if enc_type == "h":
        if priv_key:
            private_key = load_private_key(priv_key)
            if not private_key:
                print("Invalid private key provided.")
                return 1
        else:
            print("Error: No private key provided for hybrid decryption.")
            return 1

        if cwd and os.path.isdir(cwd):
            decrypt_directory(cwd, private_key)
        elif spec_dir and os.path.isdir(spec_dir):
            decrypt_directory(spec_dir, private_key)
        elif inputFile and os.path.isfile(inputFile):
            if inputFile.endswith(".enc"):
                inputFile = inputFile[:-4]  # Remove the '.enc' part

            key_file_path = os.path.join(os.path.dirname(inputFile), "key.bin")
            key = load_key(key_file_path)
            if key is None:
                print(f"Error: Could not load the key from '{key_file_path}'.")
                return 1

            decrypt_file(inputFile + ".enc", key)
            print(f"Decrypted '{inputFile}.enc'.")
        else:
            print("Invalid input path.")
            return 1

    else:
        print("Wrong usage, type ./encrypt --help for more info.")
        sys.exit(1)

    print("File(s) decrypted successfully.")
    return 0


if __name__ == "__main__":
    main()
